### ISSUE TITLE
[WIP] Long pause/hang (and/or multi-seconds per frame physics) on startup with bullet 3D physics & softbody regression

### ISSUE THREAD

**Issue by follower** [godotengine/godot#40680]
*[WIP: Started this report in case this isn't just limited to me, due to the significant performance/initial impression regression.]*

**Godot version:**

 * Not present in: Godot Engine v3.2.2.stable.official

 * Occurs in: Godot Engine v3.2.3.beta1.official

 * Still occurs in: Godot Engine v3.2.3.rc1.official

**OS/device including version:**

 * MacBook Pro mid-2012

 * OS: Mac OS X 10.9 *cough*

 * Very noticeable in GLES2 but sometimes intermittent. (Seems to occur in less severe form with GLES3.)

 * GPU: NVIDIA GeForce GT 650M OpenGL Engine

*Update 1:*

 * Issue *not* present on a different machine with:

     * GNU/Linux x86_64 Elementary OS

     * OpenGL ES 2.0 Renderer: GeForce GTX 650 Ti/PCIe/SSE2

*Update 2:*

 * Issue *not* present on:

     * iMac 2008 10.11.6

     * Godot Engine v3.2.3.beta1.official & v3.2.3.rc1.official

     * OpenGL ES 2.0 Renderer: ATI Radeon HD 2600 PRO OpenGL Engine

    However I also noticed that despite being an even older machine with older GPU & a HDD the project launched *significantly* faster with 3.2.3.rc1 than 3.2.2 did on the MacBook Pro.

    So, it's starting to look like this might be a GPU/driver issue on the original machine that was exacerbated by some subsequent change. From some of my investigation and reading some related issues it could be shader compilation related maybe?

    (I'd wondered for a while why Godot seemed to hang at startup on this machine so driver/GPU issue might be why.)

**Issue description:**

When running a project, at startup the Mac "spinning beach ball" unresponsive application indicator will appear (while showing splash screen before 3D view appears).

In addition the softbody physics will "stutter"/"jump"/"skip" and take multiple seconds per frame so the softbody jumps between positions/shapes.

This doesn't appear to be accompanied by high CPU usage but does seem to potentially be related to waiting for the GPU?

[WIP]

**Steps to reproduce:**

I haven't reduced it to a minimal test case yet--but I'd suggest either a 3D scene or a 3D scene with a SoftBody plane attached to a CSG cylinder like a flag.

I've thrown a bunch of dev tools at it but I'm still fairly in the dark.

The initial hang appears to occur after these two points:

```
1: name = 'Main::setup2', locations = 1 Options: disabled 
  1.1: where = Godot`Main::setup2(unsigned long long), address = 0x0000000100035d60, unresolved, hit count = 2 

2: name = 'OS_OSX::run', locations = 1 Options: disabled 
  2.1: where = Godot`OS_OSX::run(), address = 0x0000000100010430, unresolved, hit count = 3 
```

And between the first two calls to these points:

```
3: name = 'Main::iteration', locations = 1, resolved = 1, hit count = 33
  3.1: where = Godot`Main::iteration(), address = 0x0000000100042890, resolved, hit count = 33 

4: name = 'VisualServer::is_render_loop_enabled', locations = 1, resolved = 1, hit count = 1
  4.1: where = Godot`VisualServer::is_render_loop_enabled() const, address = 0x00000001021ef540, resolved, hit count = 1 
```

The stuttering in particular seems intermittent--from the IDE it even *seemed* like it sometimes made a difference whether I ran the scene from the toolbar or with the keyboard shortcut.

[WIP]

**Minimal reproduction project:**

[WIP]

#### Semi-related links:

 * https://github.com/godotengine/godot/issues/33834

**Comment by follower** [comment]
While investigating this further I used the Apple "Open GL Profiler" and captured this timing information during project start-up (can't remember the exact condition currently): [wacc-viewer--godot-rc1-trace.txt](https://github.com/godotengine/godot/files/4977559/wacc-viewer--godot-rc1-trace.txt)

With some sophisticated filtering (a.k.a `grep -v "   "` :) ) I highlighted the following functions which took longer than others:

<details>

<summary> <i>Expand for details of gl functions with runtime &gt; 999 µs </i> </summary>

```
0x01022727 30163.27 µs glGenerateMipmap(GL_TEXTURE_2D);
0x01022727  2554.55 µs glGenerateMipmap(GL_TEXTURE_2D);
0x01022727  2825.99 µs glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, 8, 8, 0, GL_RGB, GL_UNSIGNED_BYTE, 0x7fff58f8d9c0);
0x01022727  1857.77 µs glGenerateMipmap(GL_TEXTURE_2D);
0x01022727  1179.90 µs glGenerateMipmap(GL_TEXTURE_2D);
0x01022727 13924.50 µs glCompileShader(2);
0x01022727  2783.35 µs glCompileShader(3);
0x01022727  1029.07 µs glLinkProgram(1);
0x01022727  3382.68 µs glCheckFramebufferStatus(GL_FRAMEBUFFER); returns: GL_FRAMEBUFFER_COMPLETE 
0x01022727  3731.38 µs glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, 935, 1225, 0, GL_RGBA, GL_UNSIGNED_BYTE, 0x116e11010);
0x01022727  2021.15 µs glDrawArrays(GL_TRIANGLE_FAN, 0, 4);
0x01022727  4090.58 µs glDeleteTextures(1, {8});
0x01022727  4445.11 µs glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, 32, 16, 0, GL_RGBA, GL_UNSIGNED_BYTE, 0x7fa4ef001810);
0x01022727  2127.96 µs glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, 64, 32, 0, GL_RGBA, GL_UNSIGNED_BYTE, 0x7fa4ef07f610);
0x01022727  3897.45 µs glTexImage2D(GL_TEXTURE_2D, 0, GL_LUMINANCE_ALPHA, 16, 16, 0, GL_LUMINANCE_ALPHA, GL_UNSIGNED_BYTE, 0x7fa4eb647c80);
0x01022727  5869.91 µs glTexImage2D(GL_TEXTURE_2D, 0, GL_LUMINANCE_ALPHA, 16, 16, 0, GL_LUMINANCE_ALPHA, GL_UNSIGNED_BYTE, 0x7fa4eb63af10);
0x01022727  1004.90 µs glTexImage2D(GL_TEXTURE_2D, 0, GL_LUMINANCE_ALPHA, 16, 16, 0, GL_LUMINANCE_ALPHA, GL_UNSIGNED_BYTE, 0x7fa4eb648c10);
0x01022727  5640.43 µs glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, 16, 16, 0, GL_RGB, GL_UNSIGNED_BYTE, 0x7fa4eb6538c0);
0x01022727  5457.05 µs glTexImage2D(GL_TEXTURE_2D, 0, GL_LUMINANCE_ALPHA, 16, 16, 0, GL_LUMINANCE_ALPHA, GL_UNSIGNED_BYTE, 0x7fa4edb224e0);
0x01022727  5976.77 µs glTexImage2D(GL_TEXTURE_2D, 0, GL_LUMINANCE, 16, 16, 0, GL_LUMINANCE, GL_UNSIGNED_BYTE, 0x7fa4f15d6790);
0x01022727  2227.55 µs glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, 16, 16, 0, GL_RGBA, GL_UNSIGNED_BYTE, 0x7fa4ec89fa10);
0x01022727  2658.43 µs glClear(GL_DEPTH_BUFFER_BIT);
0x01022727 12215.31 µs glCheckFramebufferStatus(GL_FRAMEBUFFER); returns: GL_FRAMEBUFFER_COMPLETE 
0x01022727  2052.82 µs glCheckFramebufferStatus(GL_FRAMEBUFFER); returns: GL_FRAMEBUFFER_COMPLETE 
0x01022727  1760.64 µs glCheckFramebufferStatus(GL_FRAMEBUFFER); returns: GL_FRAMEBUFFER_COMPLETE 
0x01022727  1247.86 µs glCheckFramebufferStatus(GL_FRAMEBUFFER); returns: GL_FRAMEBUFFER_COMPLETE 
0x01022727  3367.96 µs glCheckFramebufferStatus(GL_FRAMEBUFFER); returns: GL_FRAMEBUFFER_COMPLETE 
0x01022727  6556.45 µs glClear(GL_COLOR_BUFFER_BIT);
0x01022727  3878.11 µs glClear(GL_DEPTH_BUFFER_BIT);
0x01022727  5325.30 µs glDeleteTextures(1, {138});
0x01022727 37522.04 µs glDeleteTextures(1, {150});
0x01022727  4795.27 µs glDeleteTextures(1, {151});
0x01022727  1083.63 µs glCheckFramebufferStatus(GL_FRAMEBUFFER); returns: GL_FRAMEBUFFER_COMPLETE 
0x01022727  1836.38 µs glCheckFramebufferStatus(GL_FRAMEBUFFER); returns: GL_FRAMEBUFFER_COMPLETE 
0x01022727  1999.99 µs glCheckFramebufferStatus(GL_FRAMEBUFFER); returns: GL_FRAMEBUFFER_COMPLETE 
0x01022727  1328.02 µs glCheckFramebufferStatus(GL_FRAMEBUFFER); returns: GL_FRAMEBUFFER_COMPLETE 
0x01022727  8979.95 µs glDeleteRenderbuffers(1, {3});
0x01022727  7267.79 µs glDeleteTextures(1, {134});
0x01022727  1444.63 µs glDeleteTextures(1, {151});
0x01022727  1484.53 µs glCheckFramebufferStatus(GL_FRAMEBUFFER); returns: GL_FRAMEBUFFER_COMPLETE 
0x01022727 16603.92 µs glCheckFramebufferStatus(GL_FRAMEBUFFER); returns: GL_FRAMEBUFFER_COMPLETE 
0x01022727  4207.38 µs glDeleteRenderbuffers(1, {3});
0x01022727  8329.89 µs glDeleteTextures(1, {138});
0x01022727  2000.94 µs glDeleteTextures(1, {139});
0x01022727  4542.85 µs glDeleteTextures(1, {140});
0x01022727  6249.69 µs glDeleteTextures(1, {141});
0x01022727  1320.00 µs glDeleteTextures(1, {151});
0x01022727  2221.37 µs glCheckFramebufferStatus(GL_FRAMEBUFFER); returns: GL_FRAMEBUFFER_COMPLETE 
0x01022727  9152.52 µs glCheckFramebufferStatus(GL_FRAMEBUFFER); returns: GL_FRAMEBUFFER_COMPLETE 
0x01022727 27706.82 µs glClear(GL_COLOR_BUFFER_BIT);
0x01022727  6507.93 µs glCheckFramebufferStatus(GL_FRAMEBUFFER); returns: GL_FRAMEBUFFER_COMPLETE 
0x01022727  4112.32 µs glCheckFramebufferStatus(GL_FRAMEBUFFER); returns: GL_FRAMEBUFFER_COMPLETE 
0x01022727  4161.65 µs glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, 1024, 512, 0, GL_RGB, GL_UNSIGNED_BYTE, 0x114095010);
0x01022727  1172.96 µs glGenerateMipmap(GL_TEXTURE_CUBE_MAP);
```
</details>

Many of which seem to meet the "GPU-CPU Sync Points" cautioned about here (in relation to WebGL): https://emscripten.org/docs/optimizing/Optimizing-WebGL.html#avoid-gpu-cpu-sync-points

But, it seems, this is probably not a general issue so probably not a huge priority for most people.

I might take a look again at some other point when I get tired of the stalls. :)

**Comment by follower** [comment]
Lol. So... It seems the stall is related to...whether or not the icon is in the dock. \*facepalm\* (*Possibly* custom icon-related?)

I was working on reducing the test case and kept on removing nodes/resources then moved to just testing a single render iteration with:

    Godot.app/Contents/MacOS/Godot --audio-driver Dummy --video-driver GLES2 --path . --quit

Which was when I noticed that v3.2.3 was *really* slow *closing* as well--and then noticed how long the icon hung around in the dock.

Previously I had v3.2.2 in the Dock (via "Keep in Dock" dock/application menu option) and when I removed it from the dock, sure it enough it stalled while starting too...

(Unfortunately because of how Godot is implemented it seems like it's not possible to have an icons kept in the dock for the Project Manager, Editor & Project all at once, so at least 2/3 will be slow.)

So... Looks like this is (a) unrelated to anything in the beta; and, (b) I suspect (older) OS version/GPU specific so will close for now.
